pkg/util/limiter/query_limiter.go

```diff
@@ -0,0 +1,78 @@
+package limiter
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/prometheus/common/model"
+
+	"github.com/cortexproject/cortex/pkg/cortexpb"
+	"github.com/cortexproject/cortex/pkg/ingester/client"
+	"github.com/cortexproject/cortex/pkg/util/validation"
+)
+
+type queryLimiterCtxKey struct{}
+
+var (
+	qlCtxKey        = &queryLimiterCtxKey{}
+	errMaxSeriesHit = "The query hit the max number of series limit while fetching chunks (limit: %d)"
```

Comment:
[nit] I would further simplify the error message. The final user has no idea what a chunk is. We should just say that the query hit the max number of series limit.

```suggestion
	errMaxSeriesHit = "The query hit the max number of series limit (limit: %d)"
```